File: app/src/main/java/com/example/core/parser/ConfigParser.kt
package com.example.core.parser

import android.util.Base64
import com.example.data.entity.ServerEntity
import org.json.JSONObject
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

/**
 * Share-link parser covering every format that mainstream clients emit.
 *
 * Supported schemes:
 *  vless, vmess (base64-JSON and plain URI), trojan, ss, ssr, hysteria,
 *  hysteria2/hy2, tuic, anytls, shadowtls, socks/socks5, http/https,
 *  wireguard/wg, ssh
 *
 * Everything is normalised into [ServerEntity]. Fields that have no dedicated
 * column are packed into existing ones with a fixed convention:
 *  - `fingerprint` doubles as the Shadowsocks cipher
 *  - `path` doubles as the TUIC password / Hysteria2 obfs password /
 *    WireGuard local address
 *  - `host` doubles as the SSH and SOCKS/HTTP username
 */
object ConfigParser {

    fun parseContent(rawContent: String, subscriptionId: Long? = null): List<ServerEntity> {
        val trimmed = rawContent.trim()
        if (trimmed.isEmpty()) return emptyList()

        val decoded = tryDecodeBase64(trimmed) ?: trimmed
        val servers = mutableListOf<ServerEntity>()

        for (line in decoded.lines()) {
            val cleanLine = line.trim()
            if (cleanLine.isBlank() || cleanLine.startsWith("#") || cleanLine.startsWith("//")) continue
            parseSingleLink(cleanLine, subscriptionId)?.let { servers.add(it) }
        }

        return servers
    }

    fun parseSingleLink(rawUri: String, subscriptionId: Long? = null): ServerEntity? {
        val t = rawUri.trim()
        val scheme = t.substringBefore("://", "").lowercase()
        return try {
            when (scheme) {
                "vless" -> parseVless(t, subscriptionId)
                "vmess" -> parseVmess(t, subscriptionId)
                "trojan", "trojan-go" -> parseTrojan(t, subscriptionId)
                "ss" -> parseShadowsocks(t, subscriptionId)
                "ssr" -> parseShadowsocksR(t, subscriptionId)
                "hysteria2", "hy2" -> parseHysteria2(t, subscriptionId)
                "hysteria", "hy" -> parseHysteria1(t, subscriptionId)
                "tuic" -> parseTuic(t, subscriptionId)
                "anytls" -> parseAnyTls(t, subscriptionId)
                "shadowtls" -> parseShadowTls(t, subscriptionId)
                "socks", "socks5" -> parseSocksOrHttp(t, "socks", subscriptionId)
                "http", "https" -> parseSocksOrHttp(t, "http", subscriptionId)
                "wireguard", "wg" -> parseWireGuard(t, subscriptionId)
                "ssh" -> parseSsh(t, subscriptionId)
                else -> null
            }
        } catch (e: Exception) {
            null
        }
    }

    // ------------------------------------------------------------- v2ray

    private fun parseVless(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host?.trim('[', ']') ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")
        val name = decodeFragment(uri.rawFragment ?: "")

        return ServerEntity(
            name = name.ifBlank { "VLESS - $host:$port" },
            protocol = "vless",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),
            security = normaliseSecurity(q["security"]),
            sni = q["sni"] ?: q["peer"] ?: "",
            publicKey = q["pbk"] ?: "",
            shortId = q["sid"] ?: "",
            fingerprint = q["fp"] ?: "chrome",
            network = normaliseNetwork(q["type"]),
            path = q["path"] ?: q["serviceName"] ?: "",
            host = q["host"] ?: "",
            alpn = q["alpn"] ?: "",
            insecure = isTrue(q["allowInsecure"]) || isTrue(q["insecure"]),
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseVmess(uriStr: String, subId: Long?): ServerEntity? {
        val payload = uriStr.substringAfter("vmess://").trim()

        // Plain-URI variant used by a few panels: vmess://uuid@host:port?...
        if (payload.contains("@")) {
            val uri = URI(uriStr)
            val host = uri.host ?: return null
            val port = if (uri.port > 0) uri.port else 443
            val q = parseQueryParams(uri.rawQuery ?: "")
            return ServerEntity(
                name = decodeFragment(uri.rawFragment ?: "").ifBlank { "VMess - $host:$port" },
                protocol = "vmess",
                server = host,
                port = port,
                uuid = uri.userInfo ?: "",
                security = normaliseSecurity(q["security"] ?: q["tls"]),
                sni = q["sni"] ?: "",
                network = normaliseNetwork(q["type"]),
                path = q["path"] ?: "",
                host = q["host"] ?: "",
                alpn = q["alpn"] ?: "",
                subscriptionId = subId,
                rawUri = uriStr
            )
        }

        val jsonStr = decodeBase64Safe(payload) ?: return null
        val json = JSONObject(jsonStr)

        val host = json.optString("add", "")
        if (host.isBlank()) return null
        val port = json.optString("port", "443").toIntOrNull() ?: 443
        val hostHeader = json.optString("host", "")
        val tls = json.optString("tls", "")
        val net = json.optString("net", "tcp")

        return ServerEntity(
            name = json.optString("ps", "").ifBlank { "VMess - $host:$port" },
            protocol = "vmess",
            server = host,
            port = port,
            uuid = json.optString("id", ""),
            alterId = json.optString("aid", "0").toIntOrNull() ?: 0,
            security = normaliseSecurity(tls),
            sni = json.optString("sni", "").ifBlank { hostHeader },
            fingerprint = json.optString("fp", "chrome"),
            network = normaliseNetwork(net),
            path = json.optString("path", "").ifBlank { json.optString("serviceName", "") },
            host = hostHeader,
            alpn = json.optString("alpn", ""),
            insecure = json.optString("allowInsecure", "") == "true",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseTrojan(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "Trojan - $host:$port" },
            protocol = "trojan",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),
            security = "tls",
            sni = q["sni"] ?: q["peer"] ?: "",
            fingerprint = q["fp"] ?: "chrome",
            network = normaliseNetwork(q["type"]),
            path = q["path"] ?: q["serviceName"] ?: "",
            host = q["host"] ?: "",
            alpn = q["alpn"] ?: "",
            insecure = isTrue(q["allowInsecure"]) || isTrue(q["insecure"]),
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    // -------------------------------------------------------- shadowsocks

    private fun parseShadowsocks(uriStr: String, subId: Long?): ServerEntity? {
        val name = decodeFragment(uriStr.substringAfter("#", ""))
        var body = uriStr.substringBefore("#").substringAfter("ss://")
        val query = if (body.contains("?")) body.substringAfter("?") else ""
        body = body.substringBefore("?")

        var method = "chacha20-ietf-poly1305"
        var password = ""
        var host: String
        var port: Int

        if (body.contains("@")) {
            val userPart = body.substringBeforeLast("@")
            val hostPart = body.substringAfterLast("@").trimEnd('/')
            val decodedUser = decodeBase64Safe(userPart) ?: URLDecoder.decode(userPart, "UTF-8")
            if (decodedUser.contains(":")) {
                method = decodedUser.substringBefore(":")
                password = decodedUser.substringAfter(":")
            } else {
                password = decodedUser
            }
            host = hostPart.substringBeforeLast(":").trim('[', ']')
            port = hostPart.substringAfterLast(":").toIntOrNull() ?: 8388
        } else {
            val decodedAll = decodeBase64Safe(body) ?: return null
            if (!decodedAll.contains("@")) return null
            val userPart = decodedAll.substringBeforeLast("@")
            val hostPart = decodedAll.substringAfterLast("@")
            method = userPart.substringBefore(":")
            password = userPart.substringAfter(":")
            host = hostPart.substringBeforeLast(":").trim('[', ']')
            port = hostPart.substringAfterLast(":").toIntOrNull() ?: 8388
        }

        if (host.isBlank()) return null
        val q = parseQueryParams(query)

        return ServerEntity(
            name = name.ifBlank { "Shadowsocks - $host:$port" },
            protocol = "shadowsocks",
            server = host,
            port = port,
            uuid = password,
            fingerprint = method,
            path = q["plugin"] ?: "",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    /** SSR links are a fully base64 blob: host:port:proto:method:obfs:base64pass/?params */
    private fun parseShadowsocksR(uriStr: String, subId: Long?): ServerEntity? {
        val blob = decodeBase64Safe(uriStr.substringAfter("ssr://").trim()) ?: return null
        val main = blob.substringBefore("/?")
        val parts = main.split(":")
        if (parts.size < 6) return null

        val host = parts[0]
        val port = parts[1].toIntOrNull() ?: return null
        val method = parts[3]
        val password = decodeBase64Safe(parts[5]) ?: parts[5]

        val q = parseQueryParams(blob.substringAfter("/?", ""))
        val remarks = q["remarks"]?.let { decodeBase64Safe(it) } ?: ""

        return ServerEntity(
            name = remarks.ifBlank { "SSR - $host:$port" },
            protocol = "shadowsocks",
            server = host,
            port = port,
            uuid = password,
            fingerprint = method,
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    // ------------------------------------------------------------- quic

    private fun parseHysteria2(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "Hysteria2 - $host:$port" },
            protocol = "hysteria2",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),
            security = "tls",
            sni = q["sni"] ?: q["peer"] ?: "",
            path = q["obfs-password"] ?: "",
            alpn = q["alpn"] ?: "h3",
            insecure = isTrue(q["insecure"]) || isTrue(q["allowInsecure"]),
            network = "quic",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseHysteria1(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "Hysteria - $host:$port" },
            protocol = "hysteria",
            server = host,
            port = port,
            uuid = q["auth"] ?: decodeFragment(uri.userInfo ?: ""),
            security = "tls",
            sni = q["peer"] ?: q["sni"] ?: "",
            alpn = q["alpn"] ?: "hysteria",
            insecure = isTrue(q["insecure"]),
            network = "quic",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseTuic(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")
        val userInfo = decodeFragment(uri.userInfo ?: "")
        val uuid = userInfo.substringBefore(":")
        val password = if (userInfo.contains(":")) userInfo.substringAfter(":") else (q["password"] ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "TUIC - $host:$port" },
            protocol = "tuic",
            server = host,
            port = port,
            uuid = uuid,
            path = password,
            security = "tls",
            sni = q["sni"] ?: "",
            alpn = q["alpn"] ?: "h3",
            insecure = isTrue(q["allow_insecure"]) || isTrue(q["insecure"]),
            network = "quic",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    // ------------------------------------------------------------- other

    private fun parseAnyTls(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "AnyTLS - $host:$port" },
            protocol = "anytls",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),
            security = "tls",
            sni = q["sni"] ?: "",
            insecure = isTrue(q["insecure"]) || isTrue(q["allowInsecure"]),
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseShadowTls(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 443
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "ShadowTLS - $host:$port" },
            protocol = "shadowtls",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),
            security = "tls",
            sni = q["sni"] ?: "",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseSocksOrHttp(uriStr: String, protocol: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else if (protocol == "socks") 1080 else 8080
        var user = ""
        var pass = ""
        val info = uri.userInfo ?: ""
        if (info.isNotBlank()) {
            val decoded = if (info.contains(":")) info else (decodeBase64Safe(info) ?: info)
            user = decoded.substringBefore(":")
            pass = if (decoded.contains(":")) decoded.substringAfter(":") else ""
        }

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { protocol.uppercase() + " - $host:$port" },
            protocol = protocol,
            server = host,
            port = port,
            uuid = pass,
            host = user,
            security = if (uriStr.startsWith("https://", true)) "tls" else "none",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseWireGuard(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 51820
        val q = parseQueryParams(uri.rawQuery ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "WireGuard - $host:$port" },
            protocol = "wireguard",
            server = host,
            port = port,
            uuid = decodeFragment(uri.userInfo ?: ""),      // private key
            publicKey = q["publickey"] ?: q["pubkey"] ?: "",
            shortId = q["presharedkey"] ?: "",
            path = q["address"] ?: q["ip"] ?: "",
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    private fun parseSsh(uriStr: String, subId: Long?): ServerEntity? {
        val uri = URI(uriStr)
        val host = uri.host ?: return null
        val port = if (uri.port > 0) uri.port else 22
        val info = decodeFragment(uri.userInfo ?: "")

        return ServerEntity(
            name = decodeFragment(uri.rawFragment ?: "").ifBlank { "SSH - $host:$port" },
            protocol = "ssh",
            server = host,
            port = port,
            uuid = if (info.contains(":")) info.substringAfter(":") else "",
            host = if (info.contains(":")) info.substringBefore(":") else info,
            subscriptionId = subId,
            rawUri = uriStr
        )
    }

    // ------------------------------------------------------------ helpers

    private fun normaliseSecurity(raw: String?): String = when (raw?.lowercase()) {
        null, "", "none", "0", "false" -> "none"
        "reality" -> "reality"
        "xtls" -> "xtls"
        else -> "tls"
    }

    private fun normaliseNetwork(raw: String?): String = when (raw?.lowercase()) {
        null, "", "tcp", "raw", "none" -> "tcp"
        "websocket" -> "ws"
        "h2", "h2c" -> "http"
        "xhttp", "splithttp" -> "xhttp"
        else -> raw.lowercase()
    }

    private fun isTrue(raw: String?): Boolean =
        raw == "1" || raw.equals("true", true) || raw.equals("yes", true)

    private fun parseQueryParams(query: String): Map<String, String> {
        if (query.isBlank()) return emptyMap()
        val result = mutableMapOf<String, String>()
        for (param in query.split("&")) {
            if (param.isBlank()) continue
            val parts = param.split("=", limit = 2)
            val key = try {
                URLDecoder.decode(parts[0], StandardCharsets.UTF_8.name())
            } catch (_: Exception) { parts[0] }
            val value = if (parts.size == 2) {
                try {
                    URLDecoder.decode(parts[1], StandardCharsets.UTF_8.name())
                } catch (_: Exception) { parts[1] }
            } else ""
            if (key.isNotBlank()) result[key.lowercase()] = value
        }
        return result
    }

    private fun decodeFragment(fragment: String): String = try {
        URLDecoder.decode(fragment, StandardCharsets.UTF_8.name())
    } catch (e: Exception) {
        fragment
    }

    private fun tryDecodeBase64(input: String): String? = try {
        val sanitized = input.replace("\r", "").replace("\n", "").trim()
        if (sanitized.contains(" ") || sanitized.contains("://")) null
        else String(
            Base64.decode(sanitized, Base64.DEFAULT or Base64.NO_WRAP or Base64.URL_SAFE),
            StandardCharsets.UTF_8
        )
    } catch (e: Exception) {
        null
    }

    private fun decodeBase64Safe(input: String): String? = try {
        val pad = when (input.length % 4) {
            2 -> "=="
            3 -> "="
            else -> ""
        }
        val clean = (input + pad).replace("-", "+").replace("_", "/")
        String(Base64.decode(clean, Base64.DEFAULT), StandardCharsets.UTF_8)
    } catch (e: Exception) {
        null
    }
}

File: windows/src/SingRay.Win/Core/ConfigParser.cs
using System.Text;
using System.Text.Json;
using SingRay.Models;

namespace SingRay.Core;

/// Parses share links and subscription payloads into ServerProfile objects.
/// Supports vless, vmess, trojan, ss, hysteria2/hy2, tuic, wireguard, socks, http,
/// plain lists and base64-encoded subscription bodies.
public static class ConfigParser
{
    private const char LF = '\u000A';
    private const char CR = '\u000D';

    public static List<ServerProfile> ParseContent(string content)
    {
        var result = new List<ServerProfile>();
        if (string.IsNullOrWhiteSpace(content)) return result;

        content = content.Trim();

        // The whole payload may be base64 (classic subscription body).
        if (!content.Contains("://") && TryDecodeBase64(content, out var decoded))
            content = decoded;

        foreach (var raw in content.Split(new[] { LF, CR }, StringSplitOptions.RemoveEmptyEntries))
        {
            var line = raw.Trim();
            if (line.Length == 0 || line.StartsWith('#')) continue;
            var profile = ParseUri(line);
            if (profile != null) result.Add(profile);
        }
        return result;
    }

    public static ServerProfile? ParseUri(string uri)
    {
        try
        {
            var scheme = uri.Split("://")[0].ToLowerInvariant();
            return scheme switch
            {
                "vless" => ParseVless(uri),
                "vmess" => ParseVmess(uri),
                "trojan" => ParseTrojan(uri),
                "ss" => ParseShadowsocks(uri),
                "hysteria2" or "hy2" => ParseHysteria2(uri),
                "tuic" => ParseTuic(uri),
                "wireguard" or "wg" => ParseWireGuard(uri),
                "socks" or "socks5" => ParseSocksOrHttp(uri, ProxyProtocol.Socks),
                _ => null
            };
        }
        catch
        {
            return null;
        }
    }

    // ------------------------------------------------------------- helpers

    private static Dictionary<string, string> Query(string query)
    {
        var dict = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (string.IsNullOrEmpty(query)) return dict;
        foreach (var pair in query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var parts = pair.Split('=', 2);
            var key = Unescape(parts[0]);
            if (string.IsNullOrWhiteSpace(key)) continue;
            dict[key] = parts.Length > 1 ? Unescape(parts[1]) : string.Empty;
        }
        return dict;
    }

    private static string Unescape(string value)
    {
        try { return Uri.UnescapeDataString(value); } catch { return value; }
    }

    private static string Fragment(string uri)
    {
        var idx = uri.IndexOf('#');
        return idx >= 0 && idx < uri.Length - 1 ? Unescape(uri.Substring(idx + 1)) : string.Empty;
    }

    private static bool TryDecodeBase64(string input, out string output)
    {
        output = string.Empty;
        try
        {
            var s = input.Trim().Replace('-', '+').Replace('_', '/');
            switch (s.Length % 4)
            {
                case 2: s += "=="; break;
                case 3: s += "="; break;
                case 1: return false;
            }
            output = Encoding.UTF8.GetString(Convert.FromBase64String(s));
            return output.Length > 0;
        }
        catch
        {
            return false;
        }
    }

    private static void ApplyStream(ServerProfile p, Dictionary<string, string> q)
    {
        p.Network = q.GetValueOrDefault("type", q.GetValueOrDefault("net", "tcp"));
        p.Security = q.GetValueOrDefault("security", p.Security);
        p.Sni = q.GetValueOrDefault("sni", q.GetValueOrDefault("peer", string.Empty));
        p.Host = q.GetValueOrDefault("host", string.Empty);
        p.Path = q.GetValueOrDefault("path", q.GetValueOrDefault("serviceName", string.Empty));
        p.Alpn = q.GetValueOrDefault("alpn", string.Empty);
        p.Flow = q.GetValueOrDefault("flow", string.Empty);
        p.Fingerprint = q.GetValueOrDefault("fp", string.Empty);
        p.PublicKey = q.GetValueOrDefault("pbk", string.Empty);
        p.ShortId = q.GetValueOrDefault("sid", string.Empty);
        var insecure = q.GetValueOrDefault("allowInsecure", q.GetValueOrDefault("insecure", "0"));
        p.AllowInsecure = insecure is "1" or "true" or "True";
        if (string.IsNullOrEmpty(p.Security) && !string.IsNullOrEmpty(p.PublicKey)) p.Security = "reality";
        if (string.IsNullOrEmpty(p.Security)) p.Security = "none";
    }

    // ------------------------------------------------------------- parsers

    private static ServerProfile ParseVless(string uri)
    {
        var u = new Uri(uri);
        var q = Query(u.Query);
        var p = new ServerProfile
        {
            Protocol = ProxyProtocol.Vless,
            Credential = u.UserInfo,
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 443,
            Name = Fragment(uri),
            RawUri = uri
        };
        ApplyStream(p, q);
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }

    private static ServerProfile? ParseVmess(string uri)
    {
        var payload = uri.Substring("vmess://".Length);

        // Standard v2rayN base64 JSON form.
        if (TryDecodeBase64(payload, out var json) && json.TrimStart().StartsWith('{'))
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            string Get(string name, string fallback = "")
            {
                if (!root.TryGetProperty(name, out var v)) return fallback;
                return v.ValueKind == JsonValueKind.Number ? v.ToString() : (v.GetString() ?? fallback);
            }

            var tls = Get("tls");
            var p = new ServerProfile
            {
                Protocol = ProxyProtocol.Vmess,
                Name = Get("ps"),
                Address = Get("add"),
                Port = int.TryParse(Get("port", "443"), out var port) ? port : 443,
                Credential = Get("id"),
                AlterId = int.TryParse(Get("aid", "0"), out var aid) ? aid : 0,
                Network = Get("net", "tcp"),
                Host = Get("host"),
                Path = Get("path"),
                Sni = Get("sni"),
                Alpn = Get("alpn"),
                Fingerprint = Get("fp"),
                Security = string.IsNullOrEmpty(tls) ? "none" : tls,
                RawUri = uri
            };
            if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
            return p;
        }

        try
        {
            var u = new Uri(uri);
            var q = Query(u.Query);
            var p = new ServerProfile
            {
                Protocol = ProxyProtocol.Vmess,
                Credential = u.UserInfo,
                Address = u.Host,
                Port = u.Port > 0 ? u.Port : 443,
                Name = Fragment(uri),
                RawUri = uri
            };
            ApplyStream(p, q);
            if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
            return p;
        }
        catch
        {
            return null;
        }
    }

    private static ServerProfile ParseTrojan(string uri)
    {
        var u = new Uri(uri);
        var q = Query(u.Query);
        var p = new ServerProfile
        {
            Protocol = ProxyProtocol.Trojan,
            Credential = Unescape(u.UserInfo),
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 443,
            Name = Fragment(uri),
            Security = "tls",
            RawUri = uri
        };
        ApplyStream(p, q);
        if (p.Security == "none") p.Security = "tls";
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }

    private static ServerProfile? ParseShadowsocks(string uri)
    {
        var body = uri.Substring("ss://".Length);
        var name = Fragment(uri);
        var hashIdx = body.IndexOf('#');
        if (hashIdx >= 0) body = body.Substring(0, hashIdx);

        var query = string.Empty;
        var qIdx = body.IndexOf('?');
        if (qIdx >= 0)
        {
            query = body.Substring(qIdx);
            body = body.Substring(0, qIdx);
        }

        string method, password, host;
        int port;

        if (body.Contains('@'))
        {
            var at = body.LastIndexOf('@');
            var userInfo = body.Substring(0, at);
            var hostPart = body.Substring(at + 1);
            if (!userInfo.Contains(':') && TryDecodeBase64(userInfo, out var decodedUser))
                userInfo = decodedUser;
            var mp = userInfo.Split(':', 2);
            method = Unescape(mp[0]);
            password = mp.Length > 1 ? Unescape(mp[1]) : string.Empty;
            var hp = hostPart.Split(':');
            host = hp[0];
            port = hp.Length > 1 && int.TryParse(hp[1], out var pp) ? pp : 8388;
        }
        else
        {
            if (!TryDecodeBase64(body, out var plain)) return null;
            var at = plain.LastIndexOf('@');
            if (at < 0) return null;
            var mp = plain.Substring(0, at).Split(':', 2);
            method = mp[0];
            password = mp.Length > 1 ? mp[1] : string.Empty;
            var hp = plain.Substring(at + 1).Split(':');
            host = hp[0];
            port = hp.Length > 1 && int.TryParse(hp[1], out var pp) ? pp : 8388;
        }

        var q = Query(query);
        return new ServerProfile
        {
            Protocol = ProxyProtocol.Shadowsocks,
            Method = method,
            Credential = password,
            Address = host,
            Port = port,
            Name = string.IsNullOrWhiteSpace(name) ? host + ":" + port : name,
            Network = q.GetValueOrDefault("type", "tcp"),
            Path = q.GetValueOrDefault("path", string.Empty),
            Host = q.GetValueOrDefault("host", string.Empty),
            RawUri = uri
        };
    }

    private static ServerProfile ParseHysteria2(string uri)
    {
        var u = new Uri(uri);
        var q = Query(u.Query);
        var p = new ServerProfile
        {
            Protocol = ProxyProtocol.Hysteria2,
            Credential = Unescape(u.UserInfo),
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 443,
            Name = Fragment(uri),
            Security = "tls",
            Sni = q.GetValueOrDefault("sni", string.Empty),
            AllowInsecure = q.GetValueOrDefault("insecure", "0") is "1" or "true",
            RawUri = uri
        };
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }

    private static ServerProfile ParseTuic(string uri)
    {
        var u = new Uri(uri);
        var q = Query(u.Query);
        var info = Unescape(u.UserInfo).Split(':', 2);
        var p = new ServerProfile
        {
            Protocol = ProxyProtocol.Tuic,
            PublicKey = info[0],
            Credential = info.Length > 1 ? info[1] : string.Empty,
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 443,
            Name = Fragment(uri),
            Security = "tls",
            Sni = q.GetValueOrDefault("sni", string.Empty),
            Alpn = q.GetValueOrDefault("alpn", "h3"),
            AllowInsecure = q.GetValueOrDefault("allow_insecure", "0") is "1" or "true",
            RawUri = uri
        };
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }

    private static ServerProfile ParseWireGuard(string uri)
    {
        var u = new Uri(uri);
        var q = Query(u.Query);
        var p = new ServerProfile
        {
            Protocol = ProxyProtocol.WireGuard,
            PrivateKey = Unescape(u.UserInfo),
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 51820,
            PeerPublicKey = q.GetValueOrDefault("publickey", q.GetValueOrDefault("pbk", string.Empty)),
            LocalAddress = q.GetValueOrDefault("address", "172.16.0.2/32"),
            PresharedKey = q.GetValueOrDefault("presharedkey", string.Empty),
            Name = Fragment(uri),
            RawUri = uri
        };
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }

    private static ServerProfile ParseSocksOrHttp(string uri, ProxyProtocol protocol)
    {
        var u = new Uri(uri);
        var info = Unescape(u.UserInfo);
        if (!info.Contains(':') && TryDecodeBase64(info, out var decoded)) info = decoded;
        var parts = info.Split(':', 2);
        var p = new ServerProfile
        {
            Protocol = protocol,
            Host = parts.Length > 0 ? parts[0] : string.Empty,
            Credential = parts.Length > 1 ? parts[1] : string.Empty,
            Address = u.Host,
            Port = u.Port > 0 ? u.Port : 1080,
            Name = Fragment(uri),
            RawUri = uri
        };
        if (string.IsNullOrWhiteSpace(p.Name)) p.Name = p.Endpoint;
        return p;
    }
}
